File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.js
function solution(numbers, target) {
    let answer = 0;
    let escapeNum = numbers.length;
    
    const bfs = (idx, sum) => {
        // 1. 재귀함수 탈출조건
        // * 재귀함수의 idx 가 numbers의 길이만큼 반복호출 됐을 경우
        if(idx === escapeNum) {
            // 누적합이 target과 같다면 answer에 +1
            if(sum === target) {
                answer++;
            }
            return;
        }
        
        // 2. 누적합에 현재숫자 + call / - call 하는 함수 호출
        bfs(idx + 1, sum + numbers[idx]); // +
        bfs(idx + 1, sum - numbers[idx]); // -
    }
    
    bfs(0,0);
    
    return answer;
}

File: 프로그래머스/2/43165. 타겟 넘버/타겟 넘버.java
class Solution {
    public int solution(int[] numbers, int target) {
        // 1. DFS 재귀 시작
        return dfs(numbers, target, 0, 0);
    }
    
    private int dfs(int[] arr, int target, int idx, int sum) {
        // 2. 하위노드 끝까지 간 경우, sum값이랑 target값 일치 비교
        if(idx == arr.length) {
            return sum == target ? 1 : 0;
        }
        
        // 3. 하위노드로 뻗어나가기 (+, -)방식으로 나뉨 => 모든 노드 탐색
        return dfs(arr, target, idx+1, sum+arr[idx])
            + dfs(arr, target, idx+1, sum-arr[idx]);
    }
}